File: Lab9/Lab9_4/Lab9_4/HtmlBuilder.cpp
#include "HtmlBuilder.h"

HtmlBuilder::HtmlBuilder(std::string path) :path(path) {
}

HtmlBuilder* HtmlBuilder::createHtml()
{
	html.open(path);
	html << "<html xmlns=\"http://www.w3.org/1999/xhtml \">" << std::endl;

	return this;
}

HtmlBuilder* HtmlBuilder::addHead()
{
	html << "<head>" << std::endl;
	html << "<title>Reporting Page</title>" << std::endl;
	html << "<link rel=\"stylesheet\" href=\"style.css\">" << std::endl;
	html << "</head>" << std::endl;
	return this;
}

HtmlBuilder* HtmlBuilder::startBody()
{
	html << "<body>" << std::endl;
	return this;
}

HtmlBuilder* HtmlBuilder::createTable()
{
	html << "<table>" << std::endl;
	addTableHeading();
	return nullptr;
}

HtmlBuilder* HtmlBuilder::addResult(int matrixSize, double sequentialTime, std::vector<std::pair<double, double>> results)
{
	html << "<tr>" << std::endl;
	html << "<td>" << matrixSize << "</td>" << std::endl;
	html << "<td>" << sequentialTime << "</td>" << std::endl;

	for (auto i : results) {
		html << "<td>" << i.first << "</td>" << std::endl;
		html << "<td>" << i.second << "</td>" << std::endl;
	}

	html << "</tr>" << std::endl;

	return this;
}

HtmlBuilder* HtmlBuilder::finishTable()
{
	html << "</table>" << std::endl;
	return nullptr;
}

HtmlBuilder* HtmlBuilder::finishBody()
{
	html << "</body>" << std::endl;
	return this;
}

void HtmlBuilder::finishHtml()
{
	html << "</html>" << std::endl;
	html.close();
}

std::ofstream* HtmlBuilder::getHtml()
{
	return &html;
}

void HtmlBuilder::addTableHeading()
{
	html << "<tr>" << std::endl;
	html << "<th rowspan=\"3\">Matrix size</th>" << std::endl;
	html << "<th rowspan=\"3\">Sequential algorithm</th>" << std::endl;
	html << "<th colspan=\"4\">Parallel algorithm</th>" << std::endl;
	html << "</tr>" << std::endl;

	html << "<tr>" << std::endl;
	html << "<th colspan=\"2\">2 processes</th>" << std::endl;
	html << "<th colspan=\"2\">4 processes</th>" << std::endl;
	html << "</tr>" << std::endl;

	html << "<tr>" << std::endl;
	for (int i = 0; i < 2; i++) {
		html << "<th>Time</th>" << std::endl;
		html << "<th>Acceleration</th>" << std::endl;
	}
	html << "</tr>" << std::endl;

}

File: Lab9/Lab9_1/src/main/java/HTMLBuilder.java
import com.hp.gagawa.java.elements.*;
import javafx.util.Pair;

import java.io.*;
import java.util.List;

public class HTMLBuilder {
    private PrintWriter out;

    private Html html;
    private Body body;

    private Table table;

    HTMLBuilder() {
        try {
            File file = new File("./src/main/resources/index.html");
            out = new PrintWriter(new FileOutputStream(file));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public void build() {
        html = new Html();
        Head head = new Head();

        Title title = new Title().appendChild(new Text("Reporting Page"));
        head.appendChild(title);

        Link link = new Link().setRel("stylesheet").setHref("style.css");
        head.appendChild(link);

        html.appendChild(head);

        body = new Body();
    }

    public void createTable() {
        table = buildTable();
    }

    public void finishTable() {
        body.appendChild(table);
    }

    public void addResult(int matrixSize, double time1, List<Pair<Double, Double>> results) {
        Tr tr = new Tr();
        tr.appendChild(new Td().appendText(Integer.toString(matrixSize)));
        tr.appendChild(new Td().appendText(Double.toString(time1)));

        for (Pair<Double, Double> value : results) {
            tr.appendChild(new Td().appendText(Double.toString(value.getKey())));
            tr.appendChild(new Td().appendText(Double.toString(value.getValue())));
        }

        table.appendChild(tr);
    }

    public void finish() {
        html.appendChild(body);
        out.println(html.write());
        out.close();
    }

    private Table buildTable() {
        table = new Table();
        addHeading();
        return table;
    }

    private void addHeading() {
        Tr heading = new Tr();
        heading.appendChild(new Th().appendText("Matrix size").setRowspan("3"));
        heading.appendChild(new Th().appendText("Sequential algorithm").setRowspan("3"));
        heading.appendChild(new Th().appendText("Parallel algorithm").setColspan("4"));

        table.appendChild(heading);

        Tr processorsNumber = new Tr();
        processorsNumber.appendChild(new Th().appendText("2 processes").setColspan("2"));
        processorsNumber.appendChild(new Th().appendText("4 processes").setColspan("2"));

        table.appendChild(processorsNumber);

        Tr parameters = new Tr();

        for (int i = 0; i < 2; i++) {
            parameters.appendChild(new Th().appendText("Time"));
            parameters.appendChild(new Th().appendText("Acceleration"));
        }

        table.appendChild(parameters);
    }
}
